src/codeu/chat/server/Controller.java

```diff
@@ -26,150 +27,133 @@
+import javax.swing.plaf.nimbus.State;
+import java.sql.*;
+import java.util.Date;
+import java.text.SimpleDateFormat;
 
-      message = new Message(id, Uuid.NULL, Uuid.NULL, creationTime, author, body);
-      model.add(message);
-      LOG.info("Message added: %s", message.id);
+import codeu.chat.common.SQLFormatter;
 
-      // Find and update the previous "last" message so that it's "next" value
-      // will point to the new message.
-
-      if (Uuid.equals(foundConversation.lastMessage, Uuid.NULL)) {
+public final class Controller implements RawController, BasicController {
 
-        // The conversation has no messages in it, that's why the last message is NULL (the first
-        // message should be NULL too. Since there is no last message, then it is not possible
-        // to update the last message's "next" value.
+    private final static Logger.Log LOG = Logger.newLog(Controller.class);
 
-      } else {
-        final Message lastMessage = model.messageById().first(foundConversation.lastMessage);
-        lastMessage.next = message.id;
-      }
+    private final Model model;
+    private final DataBaseConnection connection = new DataBaseConnection();
+    private final Uuid.Generator uuidGenerator;
 
-      // If the first message points to NULL it means that the conversation was empty and that
-      // the first message should be set to the new message. Otherwise the message should
-      // not change.
+    public Controller(Uuid serverId, Model model) {
+        this.model = model;
+        this.uuidGenerator = new RandomUuidGenerator(serverId, System.currentTimeMillis());
+        if (this.model.userByText("UNAME = 'Admin'", null).isEmpty())
+            this.model.add(new User(createId(), "Admin", Time.now(), "admin"));
+    }
 
-      foundConversation.firstMessage =
-          Uuid.equals(foundConversation.firstMessage, Uuid.NULL) ?
-          message.id :
-          foundConversation.firstMessage;
+    @Override
+    public Message newMessage(Uuid author, Uuid conversation, String body) {
+        return newMessage(createId(), author, conversation, body, Time.now());
+    }
 
-      // Update the conversation to point to the new last message as it has changed.
+    @Override
+    public User newUser(String name, String password) {
+        return newUser(createId(), name, Time.now(), password);
+    }
 
-      foundConversation.lastMessage = message.id;
+    @Override
+    public Conversation newConversation(String title, Uuid owner) {
+        return newConversation(createId(), title, owner, Time.now());
+    }
 
-      if (!foundConversation.users.contains(foundUser)) {
-        foundConversation.users.add(foundUser.id);
-      }
+    @Override
+    public Message newMessage(Uuid id, Uuid author, Uuid conversation, String body, Time creationTime) {
+
+        Collection<Message> prevMessages = model.messageByTime("MNEXTID == '0' AND CONVERSATIONID = " +SQLFormatter.sqlID(conversation), "DESC");
```

Comment:
This doesn't quite seem right. You are leaking details of the model here. Why not change the signature to allow you to call this code like model.messageByTime(conversation) ? That way you could hide the SQL in the model.